### PR TITLE
Gk g0 app all gather geo

### PR DESCRIPTION
**Pull request by akashukla**
Geometry now works in parallel. I/O is done serially until gkyl_comm_array_read is updated. Gpu geometry always reads from file. Tested on stellaramd on one gpu and my local machine on cpu with 2 cores.

### CODE REVIEW

File: zero/gkyl_gk_geometry_mapc2p.h [hunk 1]
```diff
@@ -13,15 +13,17 @@
  * @param bmag function which gives |B| in computational space
  * @param bmag_ctx Context for calculating |B|
  */
-struct gk_geometry* gkyl_gk_geometry_mapc2p_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range, const struct gkyl_range* range_ext, 
-  const struct gkyl_basis* basis, evalf_t mapc2p_func, void* mapc2p_ctx, evalf_t bmag_func, void* bmag_ctx, bool use_gpu);
+struct gk_geometry* gkyl_gk_geometry_mapc2p_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range,
```
Comment (manauref) [inline zero/gkyl_gk_geometry_mapc2p.h:16]: @akashukla please update comments accordingly

File: zero/gkyl_gk_geometry_mapc2p.h [hunk 2]
```diff
@@ -13,15 +13,17 @@
...
+    const struct gkyl_basis* basis, evalf_t mapc2p_func, void* mapc2p_ctx, evalf_t bmag_func, void* bmag_ctx, bool use_gpu);
 
 /**
  * Create a new gk geometry object that lives on NV-GPU: see new() method
  * above for documentation.
  */
 
-struct gk_geometry* gkyl_gk_geometry_mapc2p_cu_dev_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range, const struct gkyl_range* range_ext, 
-  const struct gkyl_basis* basis, evalf_t mapc2p_func, void* mapc2p_ctx, evalf_t bmag_func, void* bmag_ctx);
+struct gk_geometry* gkyl_gk_geometry_mapc2p_cu_dev_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range,
```
Comment (manauref) [inline zero/gkyl_gk_geometry_mapc2p.h:25]: @akashukla please update comments (with description of arguments) accordingly

File: zero/gkyl_gk_geometry_mirror.h [hunk 3]
```diff
@@ -11,15 +11,17 @@
  * @param mirror_rz_ctx RZ Context for use in mapping (efit file and associated quantities)
  * @param mirror_comp_ctx computation domain context for calculating mapping
  */
-struct gk_geometry* gkyl_gk_geometry_mirror_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range, const struct gkyl_range* range_ext, 
-  const struct gkyl_basis* basis, void* mirror_rz_ctx, void* mirror_comp_ctx, bool use_gpu);
+struct gk_geometry* gkyl_gk_geometry_mirror_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range,
```
Comment (manauref) [inline zero/gkyl_gk_geometry_mirror.h:14]: @akashukla please update comments accordingly

File: zero/gkyl_gk_geometry_mirror.h [hunk 4]
```diff
@@ -11,15 +11,17 @@
...
+    const struct gkyl_basis* basis, void* mirror_rz_ctx, void* mirror_comp_ctx, bool use_gpu);
 
 /**
  * Create a new gk geometry object that lives on NV-GPU: see new() method
  * above for documentation.
  */
 
-struct gk_geometry* gkyl_gk_geometry_mirror_cu_dev_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range, const struct gkyl_range* range_ext, 
-  const struct gkyl_basis* basis, void* mirror_rz_ctx, void* mirror_comp_ctx);
+struct gk_geometry* gkyl_gk_geometry_mirror_cu_dev_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range,
```
Comment (manauref) [inline zero/gkyl_gk_geometry_mirror.h:23]: @akashukla please update comments accordingly

File: zero/gkyl_gk_geometry_tok.h [hunk 5]
```diff
@@ -11,15 +11,17 @@
  * @param tok_rz_ctx RZ Context for use in mapping (efit file and associated quantities)
  * @param tok_comp_ctx computation domain context for calculating mapping
  */
-struct gk_geometry* gkyl_gk_geometry_tok_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range, const struct gkyl_range* range_ext, 
-  const struct gkyl_basis* basis, void* tok_rz_ctx, void* tok_comp_ctx, bool use_gpu);
+struct gk_geometry* gkyl_gk_geometry_tok_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range,
```
Comment (manauref) [inline zero/gkyl_gk_geometry_tok.h:14]: @akashukla please update comments accordingly

File: zero/gkyl_gk_geometry_tok.h [hunk 6]
```diff
@@ -11,15 +11,17 @@
...
+    const struct gkyl_basis* basis, void* tok_rz_ctx, void* tok_comp_ctx, bool use_gpu);
 
 /**
  * Create a new gk geometry object that lives on NV-GPU: see new() method
  * above for documentation.
  */
 
-struct gk_geometry* gkyl_gk_geometry_tok_cu_dev_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range, const struct gkyl_range* range_ext, 
-  const struct gkyl_basis* basis, void* tok_rz_ctx, void* tok_comp_ctx);
+struct gk_geometry* gkyl_gk_geometry_tok_cu_dev_new(const struct gkyl_rect_grid* grid,
```
Comment (manauref) [inline zero/gkyl_gk_geometry_tok.h:23]: @akashukla please update comments accordingly

File: zero/gk_geometry_fromfile.c [hunk 7]
```diff
@@ -14,63 +14,44 @@
 
 
 struct gk_geometry*
-gkyl_gk_geometry_fromfile_new(const struct gkyl_rect_grid* grid, const struct gkyl_range *range, const struct gkyl_range* range_ext, 
+gkyl_gk_geometry_fromfile_new(struct gk_geometry* hgeo, const struct gkyl_rect_grid* grid, const struct gkyl_range *range, const struct gkyl_range* range_ext, const struct gkyl_range *global, const struct gkyl_range* global_ext,
```
Comment (manauref) [inline zero/gk_geometry_fromfile.c:17]: @akashukla is hgeo meant to be host geo? if so please change to geo_host or geo_ho throughout, which is a convention used elsewhere

File: apps/gyrokinetic.c [hunk 8]
```diff
@@ -151,41 +152,73 @@ gkyl_gyrokinetic_app_new(struct gkyl_gk *gk)
...
     geo_basis = app->confBasis;
   }
 
   struct gk_geometry* gk_geom_3d;
   switch (gk->geometry.geometry_id) {
     case GKYL_GEOMETRY_FROMFILE:
-      gk_geom_3d = gkyl_gk_geometry_fromfile_new(&app->grid, &app->local, &app->local_ext, &app->confBasis, app->use_gpu);
+      gk_geom_3d = gkyl_gk_geometry_fromfile_new(app->gk_geom, &app->grid, &app->local, &app->local_ext, &app->global, &app->global_ext, &app->confBasis, geo_3d_use_gpu);
+      //app->gk_geom = gkyl_gk_geometry_acquire(gk_geom_3d);
+      //gkyl_gyrokinetic_app_read_geometry(app);
```
Comment (JunoRavin) [inline apps/gyrokinetic.c:183]: Can we delete these lines or is there a reason they're left in?

File: apps/gyrokinetic.c [hunk 9]
```diff
@@ -136,12 +136,13 @@ gkyl_gyrokinetic_app_new(struct gkyl_gk *gk)
...
+  struct gkyl_range geo_global_ext;
   struct gkyl_basis geo_basis;
-  bool geo_3d_use_gpu = app->use_gpu;
+  bool geo_3d_use_gpu = false;
 
   if(app->cdim < 3){
-    geo_grid = agument_grid(app->grid, gk->geometry);
-    gkyl_create_grid_ranges(&geo_grid, ghost, &geo_local_ext, &geo_local);
+    geo_grid = augment_grid(app->grid, gk->geometry);
     geo_3d_use_gpu = false;
```
Comment (JunoRavin) [inline apps/gyrokinetic.c:146]: Let's delete this line now that geo_3d_use_gpu is now always false and let's add a comment that we always initialize geometry on host.

File: zero/gk_geometry_tok.c [hunk 10]
```diff
@@ -34,65 +34,68 @@ write_nodal_coordinates(const char *nm, struct gkyl_range *nrange,
...
-    return gkyl_gk_geometry_tok_cu_dev_new(grid, range, range_ext, basis, tok_rz_ctx, tok_comp_ctx);
+    return gkyl_gk_geometry_tok_cu_dev_new(grid, range, range_ext, global, global_ext, basis, tok_rz_ctx, tok_comp_ctx);
   } 
 #endif 
 
   struct gk_geometry *up = gkyl_malloc(sizeof(struct gk_geometry));
   up->basis = *basis;
-  up->range = *range;
-  up->range_ext = *range_ext;
+  up->local = *range;
```
Comment (JunoRavin) [inline zero/gk_geometry_tok.c:49]: I think we should consistently have the inputs named local and local_ext and we should articulate this in the header files too

File: zero/gkyl_gk_geometry_mapc2p.h (hunk 2 above)
Comment (JunoRavin) [inline zero/gkyl_gk_geometry_mapc2p.h:25]: I would recommend formatting these to have the two sets of ranges on new lines (so a new line for the local and local_ext and a new line for the global and global_ext), that way you don't have any big monster lines.

**Review by JunoRavin — APPROVED** [review]

**Pull request merged**